### PR TITLE
🧪 [Frontend] 284 adding cypress e2e tests

### PR DESCRIPTION
**Pull request by xoscar**
Fixes https://github.com/open-telemetry/opentelemetry-demo/issues/284.

## Changes

- Adds E2E integration test for the frontend using Cypress.
- Fixes the backend services integration tests by dockerizing them.

https://www.loom.com/share/e6e2cbba1ca84bfb8670805b53090d77

Running e2e tests:
`docker compose run frontendtests`

Running backend integration tests:
`docker compose run integrationTests`

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -183,6 +183,49 @@ services:
       - shippingservice
     logging: *logging
 
+  # Frotend Tests
+  frontendtests:
```
Comment (xoscar) [inline docker-compose.yml:187]: Adding images for both front end cypress tests and integration tests

File: src/frontend/components/Ad/Ad.tsx [hunk 2]
```diff
@@ -7,7 +7,7 @@ const Ad = () => {
   } = useAd();
 
   return (
-    <S.Ad>
+    <S.Ad data-cy="ad">
```
Comment (xoscar) [inline src/frontend/components/Ad/Ad.tsx:10]: Adding cypress attributes

File: src/frontend/cypress.config.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+const myEnv = dotEnv.config({
+  path: resolve(__dirname, '../../.env'),
+});
+dotenvExpand.expand(myEnv);
+
+const { FRONTEND_ADDR = '', NODE_ENV, FRONTEND_PORT = '8080' } = process.env;
+
+export default defineConfig({
+  env: {
+    baseUrl: NODE_ENV === 'production' ? FRONTEND_ADDR : `http://localhost:${FRONTEND_PORT}`,
```
Comment (xoscar) [inline src/frontend/cypress.config.ts:15]: Validating the env and the port for local vs CLI execution

File: src/frontend/cypress/e2e/Checkout.cy.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+describe('Checkout Flow', () => {
+  beforeEach(() => {
+    cy.visit(Cypress.env('baseUrl'));
+  });
+
+  it('should create an order with two items', () => {
```
Comment (xoscar) [inline src/frontend/cypress/e2e/Checkout.cy.ts:6]: Some e2e tests

File: docker-compose.yml [hunk 5]
```diff
@@ -183,6 +183,50 @@ services:
...
+    environment:
+      - AD_SERVICE_ADDR
+      - CART_SERVICE_ADDR
+      - CHECKOUT_SERVICE_ADDR
+      - CURRENCY_SERVICE_ADDR
+      - PAYMENT_SERVICE_ADDR
+      - PRODUCT_CATALOG_SERVICE_ADDR
+      - RECOMMENDATION_SERVICE_ADDR
+      - SHIPPING_SERVICE_ADDR
+      - EMAIL_SERVICE_ADDR
```
Comment (mic-max) [inline docker-compose.yml:218]: nit: email in alphabetical order, same with below `depends_on`

File: src/frontend/cypress.config.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import { defineConfig } from 'cypress';
+import dotEnv from 'dotenv';
+import dotenvExpand from 'dotenv-expand';
+import { resolve } from 'path';
+
+const myEnv = dotEnv.config({
+  path: resolve(__dirname, '../../.env'),
+});
+dotenvExpand.expand(myEnv);
```
Comment (mic-max) [inline src/frontend/cypress.config.ts:9]: Oh, without this a value like `FRONTEND_ADDR=0.0.0.0:${FRONTEND_PORT}` in `.env` will remain as the literal string and not become `0.0.0.0:8080`? I didn't know this

File: src/frontend/cypress.config.ts (hunk 3 above)
Comment (mic-max) [reply src/frontend/cypress.config.ts:15]: I think instead of leaving these types of comments on the PR putting them in the source would be more valuable and they would exist in scenarios other than a code review :)

File: src/frontend/cypress/e2e/Checkout.cy.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    cy.get('[data-cy="product-card"]').first().click();
```
Comment (mic-max) [inline src/frontend/cypress/e2e/Checkout.cy.ts:7]: nit: could make a function like this to simplify the commands for reading
```javascript
const cyget(name) => cy.get(`[data-cy="${name}"]`)
```

Comment (mic-max) [reply src/frontend/cypress/e2e/Checkout.cy.ts:7]: maybe also we can define all the `data-cy` values added to the frontend code in a file and pass Enum to this `cyget` method or similar so we can avoid potential typos. ex: 

```javascript
// before
cy.get('[data-cy="cart-item-count"]')

// after
import { CART_ITEM_COUNT } from './CypressFields'
cyget(CART_ITEM_COUNT)
```

just some ideas :)

File: src/frontend/cypress/support/commands.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
+/// <reference types="cypress" />
```
Comment (mic-max) [inline src/frontend/cypress/support/commands.ts:1]: not needed?

File: src/frontend/cypress/support/e2e.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
+// ***********************************************************
```
Comment (mic-max) [inline src/frontend/cypress/support/e2e.ts:1]: also not needed?

File: test/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+FROM node:16-alpine
+
+USER node
+WORKDIR /app
+ENV NODE_ENV production
+
+COPY --chown=node:node --from=build /app/node_modules/ ./node_modules/
+COPY ./test ./
+COPY ./pb/demo.proto ./
```
Comment (mic-max) [inline test/Dockerfile:33]:
```suggestion
```

File: test/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM node:16-alpine AS build
+
+WORKDIR /app
+
+COPY ./test/package*.json ./
+
+RUN apk add --no-cache python3 make g++ && npm ci
```
Comment (mic-max) [inline test/Dockerfile:21]:
```suggestion
RUN apk add --no-cache python3 make g++ && npm ci --omit=dev
```
Is the `apk add --no-cache python3 make g++` needed? if so for what?

File: test/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+USER node
+WORKDIR /app
+ENV NODE_ENV production
+
+COPY --chown=node:node --from=build /app/node_modules/ ./node_modules/
+COPY ./test ./
+COPY ./pb/demo.proto ./
+
+ENTRYPOINT [ "sh", "-c", "npm test" ]
```
Comment (mic-max) [inline test/Dockerfile:35]:
```suggestion
ENTRYPOINT [ "npm", "test" ]
```
does this not work? what does `sh -c` do?

File: test/test.js [hunk 13]
```diff
@@ -1,243 +1,306 @@
...
 
-  const res = await currencyConvert(req)
-  t.is(res.currencyCode, "CAD")
-  t.is(res.units, 442)
-  t.is(res.nanos, 599380805)
-})
+  const res = await currencyConvert(req);
+  t.is(res.currencyCode, "CAD");
+  t.is(res.units, 442);
+  t.is(res.nanos, 599380804);
```
Comment (mic-max) [inline test/test.js:179]: this is interesting. this might be system dependent because I added this and it was passing...
This might need to be changed to allow some wiggle room on either side
```suggestion
  t.true(res.nanos >= 599380800);
  t.true(res.nanos < 599380809);
```
something like this for example..

File: test/test.js [hunk 14]
```diff
@@ -1,243 +1,306 @@
 // Node
```
Comment (mic-max) [inline test/test.js:1]: Was the formatting of this file causing some linting errors? I'm fine with the change but ideally pure refactors should be a separate PR to easily review and merge quick so less people or possibly yourself maybe be affected by conflicts.

**Review by mic-max — CHANGES_REQUESTED** [review]
LGTM other than some minor questions / feedback I left

File: docker-compose.yml [hunk 15]
```diff
@@ -183,6 +183,50 @@ services:
       - shippingservice
     logging: *logging
 
+  # Frotend Tests
```
Comment (jorgeepc) [inline docker-compose.yml:186]: typo `Frontend`

File: docker-compose.yml [hunk 16]
```diff
@@ -183,6 +183,50 @@ services:
...
+      - tests
+    environment:
+      - CYPRESS_baseUrl=http://${FRONTEND_ADDR}
+      - NODE_ENV=production
+    working_dir: /cypress
+    volumes:
+      - ./src/frontend:/cypress
+
+  # Integration Tests
+  integrationTests:
```
Comment (jorgeepc) [inline docker-compose.yml:201]: We should sync the case for `frontendtests` and `integrationTests`. Maybe `frontendTests`?

File: src/frontend/cypress.config.ts (hunk 3 above)
Comment (xoscar) [reply src/frontend/cypress.config.ts:15]: I tent do disagree, I feel that the code itself should be self-explanatory but for PRs it is good to present it as a story for the reviewer so even if they don't have enough context the basic flow of changes can be understood

File: src/frontend/cypress/e2e/Checkout.cy.ts (hunk 7 above)
Comment (xoscar) [reply src/frontend/cypress/e2e/Checkout.cy.ts:7]: Great idea!

File: test/Dockerfile (hunk 10 above)
Comment (xoscar) [reply test/Dockerfile:33]: I think we need to keep this as the tests need the proto file

File: test/test.js (hunk 14 above)
Comment (xoscar) [reply test/test.js:1]: Make sense, I just ran the prettify command and it changed it 😬

**Comment by xoscar** [comment]
Hey @mic-max thanks for the review! I just updated the PR, let me know what you think 😄

**Review by mic-max — APPROVED** [review]

**Comment by cartersocha** [comment]
Lgtm. @xoscar could you add a change log entry & I'll merge

**Comment by xoscar** [comment]
@cartersocha done 😄
